src/sentry/utils/performance_issues/performance_detection.py

```diff
@@ -1395,6 +1403,202 @@ def _is_file_io_on_main_thread(self, span: Span) -> bool:
+        found_different_span = False
+
+        for span in pattern:
+            op = span.get("op") or ""
+            description = span.get("description") or ""
+            found_db_op = found_db_op or (
+                op.startswith("db") and description and not description.endswith("...")
+            )
+            found_different_span = found_different_span or not self._equivalent(pattern[0], span)
+            if found_db_op and found_different_span:
+                return True
+
+        return False
+
+
+class ContinuingMNPlusOne(MNPlusOneState):
+    """
+    The state for when we think we might have found a pattern: a sequence of
+    spans that has begun to repeat.
+
+    When the sequence is broken (either by a mismatched span or span iteration
+    finishing), returns to the SearchingMNPlusOne state, possibly returning a
+    PerformanceProblem if the detected sequence met our thresholds.
+    """
+
+    __slots__ = ("settings", "pattern", "spans", "times_repeated", "pattern_index")
+
+    def __init__(self, settings: Dict[str, Any], pattern: Sequence[Span], first_span: Span) -> None:
+        self.settings = settings
+        self.pattern = pattern
+
+        # The full list of spans involved in the MN pattern.
+        self.spans = pattern.copy()
+        self.spans.append(first_span)
+        self.times_repeated = 0
+        self.pattern_index = 1
+
+    def next(self, span: Span) -> MNPlusOneState:
+        # If the MN pattern is continuing, carry on in this state.
+        pattern_span = self.pattern[self.pattern_index]
+        if self._equivalent(pattern_span, span):
+            self.spans.append(span)
+            self.pattern_index += 1
+            if self.pattern_index >= len(self.pattern):
+                self.times_repeated += 1
+                self.pattern_index = 0
+            return (self, None)
+
+        # We've broken the MN pattern, so return to the Searching state. If it
+        # is a significant problem, also return a PerformanceProblem.
+        self.spans.append(span)
+        start_index = len(self.pattern) * (self.times_repeated + 1)
+        return (
+            SearchingForMNPlusOne(self.settings, self.spans[start_index:]),
+            self._maybe_performance_problem(),
+        )
+
+    def finish(self) -> Optional[PerformanceProblem]:
+        return self._maybe_performance_problem()
+
+    def _maybe_performance_problem(self) -> Optional[PerformanceProblem]:
+        minimum_times_repeated = self.settings["minimum_times_repeated"]
+        if self.times_repeated < minimum_times_repeated:
+            return None
+
+        offender_span_count = len(self.pattern) * (self.times_repeated + 1)
+        offender_spans = self.spans[:offender_span_count]
+
+        total_duration_threshold = self.settings["total_duration_threshold"]
+        total_duration = functools.reduce(
+            lambda a, b: a + b, [get_span_duration(span) for span in offender_spans], timedelta(0)
+        )
```

Comment:
I seem to be more `reduce`-averse than others but this could also be
```suggestion
        total_duration = sum(map(get_span_duration, offender_spans), timedelta(0))
```

Reply:
I like yours better: shorter, and `sum` communicates what's happening better than the lambda. Awesome!